### PR TITLE
Improve cmake first time error reporting

### PR DESCRIPTION
**Pull request by daschuer**
This aims to fix https://github.com/mixxxdj/mixxx/issues/11596 and https://github.com/mixxxdj/mixxx/issues/11597

One issue is that we silently add Googletest and it is not obvious that the missing pthread.h comes from it and not from Mixxx. @ikr83 has blamed the previous check of sccach which has been already correctly installed via scoop. 

The other issue is that cmake has to find the CMAKE_TOOLCHAIN in one of the various ways. 

Now cake gives a hint how to install it and the tool scripts give a hint that you need to wipe the build directory and add the path to the toolchain. 

I have decided to use the full path and avoid environment variables. This should work independent from the selected shell and independent from how the script has been started.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1994,8 +1995,27 @@ if(WIN32)
...
+    if(CMAKE_BUILD_TYPE MATCHES "Debug")
+      message(FATAL_ERROR "Did you download the build environment via ´${CMAKE_SOURCE_DIR}/tools/windows_buildenv.bat´ and run ")
+	else()
+	  message(FATAL_ERROR "Did you download the build environment via ´${CMAKE_SOURCE_DIR}/tools/windows_release_buildenv.bat´")
+	endif()
+  elseif(APPLE)
+    if(CMAKE_BUILD_TYPE MATCHES "Debug")
+      message(FATAL_ERROR "Did you download the build environment via ´${CMAKE_SOURCE_DIR}/tools/macos_buildenv.bat´")
+	else()
+	  message(FATAL_ERROR "Did you download the build environment via ´${CMAKE_SOURCE_DIR}/tools/macos_release_buildenv.bat´")
```
Comment (JoergAtGithub) [inline CMakeLists.txt:2011]:
```suggestion
	  message(FATAL_ERROR "Did you download the build environment via ´${CMAKE_SOURCE_DIR}/tools/macos_buildenv.bat´ or ´${CMAKE_SOURCE_DIR}/tools/macos_release_buildenv.bat´")
```

File: tools/windows_buildenv.bat [hunk 2]
```diff
@@ -186,6 +186,11 @@ REM Generate CMakeSettings.json which is read by MS Visual Studio to determine t
     CALL :Configuration2CMakeSettings_JSON native    Release
     >>"%CMakeSettings%" echo   ]
     >>"%CMakeSettings%" echo }
+
+    echo ^You can now open CMakeSetting.json from Visual Studio
```
Comment (JoergAtGithub) [inline tools/windows_buildenv.bat:190]:
```suggestion
    echo ^You can now use the "Open Folder" function from Visual Studio to open the mixxx directory and the settings will be read from the included CMakeSetting.json file,
```

**Review by JoergAtGithub — CHANGES_REQUESTED** [review]
The hint doesn't appear for me, in case of missing buildenv:
```
CMake generation started for configuration: 'x64__legacy'.
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="RelWithDebInfo" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\install\x64__legacy" -DCMAKE_TOOLCHAIN_FILE:FILEPATH="C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\buildenv\mixxx-deps-2.4-x64-windows-f9db28e\scripts\buildsystems\vcpkg.cmake" -DMIXXX_VCPKG_ROOT:STRING="C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\buildenv\mixxx-deps-2.4-x64-windows-f9db28e" -DBATTERY:BOOL="True" -DBROADCAST:BOOL="True" -DBULK:BOOL="False" -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL="True" -DDEBUG_ASSERTIONS_FATAL:BOOL="True" -DFFMPEG:BOOL="False" -DHID:BOOL="True" -DHSS1394:BOOL="True" -DKEYFINDER:BOOL="False" -DLOCALECOMPARE:BOOL="True" -DLILV:BOOL="True" -DMAD:BOOL="True" -DMEDIAFOUNDATION:BOOL="True" -DMODPLUG:BOOL="True" -DOPUS:BOOL="True" -DOPTIMIZE:STRING="legacy" -DQTKEYCHAIN:BOOL="True" -DSTATIC_DEPS:BOOL="False" -DVINYLCONTROL:BOOL="True" -DWAVPACK:BOOL="True" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe"  -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "C:\Users\Joerg.WORLDWARTWEB\Source\Repos\JoergAtGithub\mixxx" 2>&1"
1> Working directory: C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\build\x64__legacy
1> [CMake] -- Using MIXXX_VCPKG_ROOT: C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\buildenv\mixxx-deps-2.4-x64-windows-f9db28e
1> [CMake] CMake Error at build/x64__legacy/CMakeFiles/3.26.0-msvc3/CMakeSystem.cmake:6 (include):
1> [CMake]   include could not find requested file:
1> [CMake] 
1> [CMake]     C:/Users/Joerg.WORLDWARTWEB/source/repos/JoergAtGithub/mixxx/buildenv/mixxx-deps-2.4-x64-windows-f9db28e/scripts/buildsystems/vcpkg.cmake
1> [CMake] Call Stack (most recent call first):
1> [CMake]   CMakeLists.txt:144 (project)
1> [CMake] CMake Error at C:\Users\Joerg.WORLDWARTWEB\Source\Repos\JoergAtGithub\mixxx\CMakeLists.txt:168 (message):
1> [CMake]   VCPKG_TARGET_TRIPLET dir not found:
1> [CMake]   C:/Users/Joerg.WORLDWARTWEB/source/repos/JoergAtGithub/mixxx/buildenv/mixxx-deps-2.4-x64-windows-f9db28e/installed/x64-windows
1> [CMake]   Make sure the VCPKG build environment is installed and contains the build
1> [CMake]   for the selected triplet.
1> [CMake] 
1> [CMake] 
1> [CMake] -- Configuring incomplete, errors occurred!
1> 'C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="RelWithDebInfo" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\install\x64__legacy" -DCMAKE_TOOLCHAIN_FILE:FILEPATH="C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\buildenv\mixxx-deps-2.4-x64-windows-f9db28e\scripts\buildsystems\vcpkg.cmake" -DMIXXX_VCPKG_ROOT:STRING="C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\buildenv\mixxx-deps-2.4-x64-windows-f9db28e" -DBATTERY:BOOL="True" -DBROADCAST:BOOL="True" -DBULK:BOOL="False" -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL="True" -DDEBUG_ASSERTIONS_FATAL:BOOL="True" -DFFMPEG:BOOL="False" -DHID:BOOL="True" -DHSS1394:BOOL="True" -DKEYFINDER:BOOL="False" -DLOCALECOMPARE:BOOL="True" -DLILV:BOOL="True" -DMAD:BOOL="True" -DMEDIAFOUNDATION:BOOL="True" -DMODPLUG:BOOL="True" -DOPUS:BOOL="True" -DOPTIMIZE:STRING="legacy" -DQTKEYCHAIN:BOOL="True" -DSTATIC_DEPS:BOOL="False" -DVINYLCONTROL:BOOL="True" -DWAVPACK:BOOL="True" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe"  -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "C:\Users\Joerg.WORLDWARTWEB\Source\Repos\JoergAtGithub\mixxx" 2>&1"' execution failed with error: ''C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="RelWithDebInfo" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\install\x64__legacy" -DCMAKE_TOOLCHAIN_FILE:FILEPATH="C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\buildenv\mixxx-deps-2.4-x64-windows-f9db28e\scripts\buildsystems\vcpkg.cmake" -DMIXXX_VCPKG_ROOT:STRING="C:\Users\Joerg.WORLDWARTWEB\source\repos\JoergAtGithub\mixxx\buildenv\mixxx-deps-2.4-x64-windows-f9db28e" -DBATTERY:BOOL="True" -DBROADCAST:BOOL="True" -DBULK:BOOL="False" -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL="True" -DDEBUG_ASSERTIONS_FATAL:BOOL="True" -DFFMPEG:BOOL="False" -DHID:BOOL="True" -DHSS1394:BOOL="True" -DKEYFINDER:BOOL="False" -DLOCALECOMPARE:BOOL="True" -DLILV:BOOL="True" -DMAD:BOOL="True" -DMEDIAFOUNDATION:BOOL="True" -DMODPLUG:BOOL="True" -DOPUS:BOOL="True" -DOPTIMIZE:STRING="legacy" -DQTKEYCHAIN:BOOL="True" -DSTATIC_DEPS:BOOL="False" -DVINYLCONTROL:BOOL="True" -DWAVPACK:BOOL="True" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe"  -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "C:\Users\Joerg.WORLDWARTWEB\Source\Repos\JoergAtGithub\mixxx" 2>&1"' returned with exit code: 1'.
```

**Comment by daschuer** [comment]
OK, message fixed and it appears now at more early places.

File: CMakeLists.txt [hunk 3]
```diff
@@ -59,7 +77,8 @@ if(DEFINED MIXXX_VCPKG_ROOT)
     # and we configure the CMAKE_TOOLCHAIN_FILE and overlays accordingly
     message(STATUS "Using MIXXX_VCPKG_ROOT: $ENV{MIXXX_VCPKG_ROOT}")
   else()
-    message(FATAL_ERROR "MIXXX_VCPKG_ROOT not correct (missing $ENV{MIXXX_VCPKG_ROOT}/overlay/ports)")
+    message(STATUS "MIXXX_VCPKG_ROOT not correct (missing $ENV{MIXXX_VCPKG_ROOT}/overlay/ports)")
```
Comment (JoergAtGithub) [inline CMakeLists.txt:80]: I wonder if the if condition is correct here. Shouldn't both directories exist here?

**Review by JoergAtGithub — CHANGES_REQUESTED** [review]
This PR contains the unrelated commit [1f9f176](https://github.com/mixxxdj/mixxx/pull/11635/commits/1f9f176fb63683b0f0c0543111bc7009720aa44d), please remove it.

File: CMakeLists.txt [hunk 4]
```diff
@@ -45,6 +45,24 @@ if(POLICY CMP0135)
   cmake_policy(SET CMP0135 NEW)
 endif()
 
+function(FATAL_ERROR_MISSING_ENV)
+  if(WIN32)
+    if(CMAKE_BUILD_TYPE MATCHES "Debug")
+      message(FATAL_ERROR "Did you download the build environment via ´${CMAKE_SOURCE_DIR}/tools/windows_buildenv.bat´?")
```
Comment (JoergAtGithub) [inline CMakeLists.txt:51]:
```suggestion
      message(FATAL_ERROR "Did you download the Mixxx build environment using ´${CMAKE_SOURCE_DIR}/tools/windows_buildenv.bat´?")
```
The suggested wording makes it maybe more clear, that the buildenv is something, that you get from the Mixxx project. What do you think?

**Comment by daschuer** [comment]
Done.

**Review by JoergAtGithub — APPROVED** [review]
Tested on Windows11 and works as expected! Thank you!

**Pull request merged**